fix: fix timezone issues with current event

dodumosu opened the pull request:
see comments on nditech/apollo-issues#41

dodumosu committed:
fix: fix timezone issues with current event

takinbo commented:
Thanks for the PR. Coincidentally, I started working on this and started rethinking the purpose for how we determine the current event.

So we currently have this:

```python
now = current_timestamp()

lower_bound = datetime.combine(now, datetime.min.time())
upper_bound = datetime.combine(now, datetime.max.time())

# convert back to UTC because datetime.combine() returns
# naive objects
lower_bound = pytz.utc.localize(lower_bound)
upper_bound = pytz.utc.localize(upper_bound)

event = self.filter(
    Event.start <= upper_bound, Event.end >= lower_bound
).order_by(Event.start.desc()).first()
```

I forget why we need to fit the current event within the current day. Why not simply use the current timestamp and compare with that instead? I mean it is possible to have several events within the day! No?

So I rewrote it as this:

```python
now = current_timestamp()

# find events overlapping with the current timestamp and return
# the most recent.
event = self.filter(
    Event.start <= now, Event.end >= now
).order_by(Event.start.desc()).first()
```

Thoughts?

dodumosu commented:
it's not usual, but it's happened. for example, i recall an instance when in Nigeria a training event (i believe it was across several states) was split across a single day. still, that could be a workflow issue.

takinbo commented:
What are the drawbacks of simplifying current event retrieval. I've done some testing with the Zambia current set of events and it appears to do what is expected. I'm trying to understand what were the considerations that led us to have a complicated system for determining what event to select as the current one.

dodumosu commented:
i don't think there are any demerits to simply using the timestamp. i guess we were just trying to get a window for active events, instead of an instant.

dodumosu commented:
so this is what it looks like now based on the discussion above:
```python
now = current_timestamp(use_app_timezone=True)

event = self.filter(
    Event.start <= now, Event.end >= now
).order_by(Event.start.desc()).first()

if event:
    return event

# if there's no event, pick the closest past event
event = self.filter(Event.end <= now).order_by(
    Event.end.desc()).first()

if event:
    return event

# if there's no event, pick the closest future event
event = self.filter(Event.start >= now).order_by(
    Event.start).first()

if event:
    return event

return None
```

takinbo commented:
``current_timestamp`` currently returns the current timestamp in UTC format. Event start and end times are stored in UTC so I don't think there's a need to use the ``use_app_timezone`` option.

dodumosu committed:
feat: change active event implementation

rather than use a day range for the active event query,
this commit uses the current instance.
it also reverts the implementation of `apollo.utils.current_timestamp`.

takinbo commented on apollo/utils.py:
```diff
@@ -28,7 +28,7 @@ def parse_env(env_path):
 
 
-def current_timestamp():
+def current_timestamp(use_app_timezone=False):
```
``use_app_timezone`` is extraneous.

takinbo commented on apollo/deployments/services.py:
```diff
@@ -1,48 +1,34 @@
-        # naive objects
-        lower_bound = pytz.utc.localize(lower_bound)
-        upper_bound = pytz.utc.localize(upper_bound)
+        now = current_timestamp(use_app_timezone=True)
```
when ``use_app_timezone`` is removed in the method definition, there should be no need for specifying it here too.

dodumosu committed:
fix: remove argument

dodumosu committed:
chore: remove unused parameter

takinbo reviewed (approved):
utACK
